pkg/agent/evictionmanager/plugin/memory/numa_pressure.go

```diff
@@ -121,51 +129,90 @@ func (n *NumaMemoryPressurePlugin) ThresholdMet(_ context.Context) (*pluginapi.T
+			}
+		} else {
+			// If there is no watermark info in zoneinfo, fall back to applying info from metaServer
+			if err := n.detectNumaWatermarkPressure(numaID, 0, 0, 0); err != nil {
+				errList = append(errList, err)
+				continue
+			}
 		}
 	}
 
 	return errors.NewAggregate(errList)
 }
 
-func (n *NumaMemoryPressurePlugin) detectNumaWatermarkPressure(numaID int) error {
-	free, total, scaleFactor, err := helper.GetWatermarkMetrics(n.metaServer.MetricsFetcher, n.emitter, numaID)
-	if err != nil {
-		general.Errorf("failed to getWatermarkMetrics for numa %d, err: %v", numaID, err)
-		_ = n.emitter.StoreInt64(metricsNameFetchMetricError, 1, metrics.MetricTypeNameCount,
-			metrics.ConvertMapToTags(map[string]string{
-				metricsTagKeyNumaID: strconv.Itoa(numaID),
-			})...)
-		return err
+func (n *NumaMemoryPressurePlugin) isUnderAdditionalPressure(free, min, low int) bool {
+	return free < (min+low)/2 || (low > memGapPages && free < (low-memGapPages))
+}
+
+func (n *NumaMemoryPressurePlugin) detectNumaWatermarkPressure(numaID, free, min, low int) error {
+	if low <= 0 {
+		// If there is no watermark info in zoneinfo, fall back to applying info from metaServer
+		// Notice: the unit of freeMem here is bytes.
+		freeBytes, totalBytes, scaleFactor, err := helper.GetWatermarkMetrics(n.metaServer.MetricsFetcher, n.emitter, numaID)
+		if err != nil {
+			general.Errorf("failed to getWatermarkMetrics for numa %d, err: %v", numaID, err)
+			_ = n.emitter.StoreInt64(metricsNameFetchMetricError, 1, metrics.MetricTypeNameCount,
+				metrics.ConvertMapToTags(map[string]string{
+					metricsTagKeyNumaID: strconv.Itoa(numaID),
+				})...)
+			return err
+		}
+		free = general.ConvertBytesToPages(int(freeBytes))
+		low = general.ConvertBytesToPages(int(totalBytes * scaleFactor / 10000))
 	}
 
 	dynamicConfig := n.dynamicConfig.GetDynamicConfiguration()
 	general.Infof("numa watermark metrics of ID: %d, "+
-		"free: %+v, total: %+v, scaleFactor: %+v, numaFreeBelowWatermarkTimes: %+v, numaFreeBelowWatermarkTimesThreshold: %+v",
-		numaID, free, total, scaleFactor, n.numaFreeBelowWatermarkTimesMap[numaID],
+		"free pages: %+v, min pages: %+v, low pages: %+v, numaFreeBelowWatermarkTimes: %+v, numaFreeBelowWatermarkTimesThreshold: %+v",
+		numaID, free, min, low, n.numaFreeBelowWatermarkTimesMap[numaID],
 		dynamicConfig.NumaFreeBelowWatermarkTimesThreshold)
 	_ = n.emitter.StoreFloat64(metricsNameNumaMetric, float64(n.numaFreeBelowWatermarkTimesMap[numaID]), metrics.MetricTypeNameRaw,
 		metrics.ConvertMapToTags(map[string]string{
 			metricsTagKeyNumaID:     strconv.Itoa(numaID),
 			metricsTagKeyMetricName: metricsTagValueNumaFreeBelowWatermarkTimes,
 		})...)
 
-	if free < total*scaleFactor/10000 {
-		n.isUnderNumaPressure = true
-		n.numaActionMap[numaID] = actionReclaimedEviction
+	if free < low {
 		n.numaFreeBelowWatermarkTimesMap[numaID]++
+		if n.numaFreeBelowWatermarkTimesMap[numaID] > (dynamicConfig.NumaFreeBelowWatermarkTimesThreshold / 2) {
```

Comment:
Why this threshold need be divided by 2

Reply:
Currently, the default value for NumaFreeBelowWatermarkTimesThreshold is 20,
 with a default periodic check interval of 5 seconds.
This means if the free memory remains below the low watermark for 50 seconds, actionReclaimedEviction will be triggered.
If the condition persists for 100 seconds, actionEviction will be triggered.